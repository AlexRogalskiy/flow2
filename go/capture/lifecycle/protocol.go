package lifecycle

import (
	"encoding/json"
	"fmt"

	pc "github.com/estuary/flow/go/protocols/capture"
	pf "github.com/estuary/flow/go/protocols/flow"
)

// StageCaptured potentially sends a previously staged Captured into the stream,
// and then stages its arguments into response.Captured.
func StageCaptured(
	stream interface {
		Send(*pc.CaptureResponse) error
	},
	response **pc.CaptureResponse,
	binding int,
	document json.RawMessage,
) error {
	// Send current |response| if we would re-allocate.
	if *response != nil {
		var rem int
		if l := (*response).Captured; int(l.Binding) != binding {
			rem = -1 // Must flush this response.
		} else if cap(l.DocsJson) != len(l.DocsJson) {
			rem = cap(l.Arena) - len(l.Arena)
		}
		if rem < len(document) {
			if err := stream.Send(*response); err != nil {
				return fmt.Errorf("sending Captured response: %w", err)
			}
			*response = nil
		}
	}

	if *response == nil {
		*response = &pc.CaptureResponse{
			Captured: &pc.CaptureResponse_Captured{
				Binding:  uint32(binding),
				Arena:    make(pf.Arena, 0, arenaSize),
				DocsJson: make([]pf.Slice, 0, sliceSize),
			},
		}
	}

	var l = (*response).Captured
	l.DocsJson = append(l.DocsJson, l.Arena.Add(document))

	return nil
}

// WriteCommit flushes a pending Captured response, and sends a Commit response
// with the provided driver checkpoint.
func WriteCommit(
	stream interface {
		Send(*pc.CaptureResponse) error
	},
	response **pc.CaptureResponse,
	commit *pc.CaptureResponse_Commit,
) error {
	// Flush partial Captured response, if required.
	if *response != nil {
		if err := stream.Send(*response); err != nil {
			return fmt.Errorf("flushing final Captured response: %w", err)
		}
		*response = nil
	}

	if err := stream.Send(&pc.CaptureResponse{
		Commit: commit,
	}); err != nil {
		return fmt.Errorf("sending Commit response: %w", err)
	}

	return nil
}

const (
	arenaSize = 16 * 1024
	sliceSize = 32
)
