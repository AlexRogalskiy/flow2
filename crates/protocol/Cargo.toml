[package]
name = "protocol"
version = "0.0.0"
authors = ["Estuary Technologies, Inc"]
edition = "2018"

[dependencies]
prost = "*"
prost-types = "*"
serde = "*"
tonic = "*"

[dev-dependencies]
insta = "*"
serde_json = "*"

[build-dependencies]
tonic-build = "*"
