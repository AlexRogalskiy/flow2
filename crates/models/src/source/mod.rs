use crate::tables;

mod loader;
pub mod scenarios;
mod scope;
pub mod specs;
pub mod wrappers;

pub use loader::{FetchResult, LoadError, Loader};
pub use scope::Scope;

pub use specs::*;
pub use wrappers::*;

#[derive(Default, Debug)]
pub struct Tables {
    pub captures: tables::Captures,
    pub collections: tables::Collections,
    pub derivations: tables::Derivations,
    pub endpoints: tables::Endpoints,
    pub errors: tables::Errors,
    pub fetches: tables::Fetches,
    pub imports: tables::Imports,
    pub materializations: tables::Materializations,
    pub nodejs_dependencies: tables::NodeJSDependencies,
    pub projections: tables::Projections,
    pub resources: tables::Resources,
    pub schema_docs: tables::SchemaDocs,
    pub test_steps: tables::TestSteps,
    pub transforms: tables::Transforms,
}

impl Tables {
    // Access Tables as an array of dynamic TableObj instances.
    pub fn as_tables(&self) -> Vec<&dyn tables::TableObj> {
        // This de-structure ensures we can't fail to update if fields change.
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }

    // Access Tables as an array of mutable dynamic TableObj instances.
    pub fn as_tables_mut(&mut self) -> Vec<&mut dyn tables::TableObj> {
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }
}
