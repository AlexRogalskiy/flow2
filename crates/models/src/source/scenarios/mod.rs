use crate::source;

use futures::channel::oneshot;
use futures::future::FutureExt;
use serde_json::Value;
use std::cell::RefCell;
use std::collections::BTreeMap;
use std::task::Poll;
use url::Url;

macro_rules! file_tests {
    ($($name:ident,)*) => {
    $(
        #[test]
        fn $name() {
			let fixture = include_bytes!(concat!(stringify!($name), ".yaml"));
            let fixture: Value = serde_yaml::from_slice(fixture).unwrap();
			let tables = evaluate_fixtures(Default::default(), &fixture);
			insta::assert_debug_snapshot!(tables);
        }
    )*
    }
}

file_tests! {
    test_catalog_import_cycles,
    test_collections,
    test_derivations,
    test_schema_with_inline,
    test_schema_with_nested_ids,
    test_schema_with_references,
    test_simple_catalog,
    test_test_case,
    test_endpoints_captures_materializations,
}

pub fn evaluate_fixtures(catalog: source::Tables, fixture: &Value) -> source::Tables {
    let fixtures = match fixture {
        Value::Object(m) => m,
        _ => panic!("fixtures must be an object having resource properties"),
    };

    // Fetches holds started fetches since the last future poll.
    // Use an ordered map so that we signal one-shots in a stable order,
    // making snapshots reliable.
    let fetches: RefCell<BTreeMap<String, oneshot::Sender<source::FetchResult>>> =
        RefCell::new(BTreeMap::new());

    // Fetch function which queues oneshot futures for started fetches.
    let fetch = |url: &Url| {
        let (tx, rx) = oneshot::channel();
        if let Some(_) = fetches.borrow_mut().insert(url.to_string(), tx) {
            panic!("url {} has already been fetched", url);
        }
        rx.map(|r| r.unwrap())
    };
    let loader = source::Loader::new(catalog, fetch);
    let root = Url::parse("test://root").unwrap();

    // What's going on here? Glad you asked.
    //
    // loader.load_resource() is returning a Future, which under the covers
    // is a big, synchronously-invoked, and deterministic state machine that's
    // built with the compiler's assistance.
    //
    // By construction, we know that there's only one real "await" point of this
    // future which will cause it to return Poll::Pending: when it's called into the
    // |fetch| closure we gave it above, and it has no work remaining to do until
    // at least one of the Futures returned by |fetch| resolves.
    //
    // Loader is walking sources concurrently. It processes fetches as separate
    // internal (still synchronous & deterministic) tasks, so we *do* expect to
    // see multiple |fetch| calls made in between Poll::Pending poll results of
    // the future.
    //
    // So, the strategy is this: we _synchronously_ poll the future forward in a
    // loop (e.x., we're not using a runtime *at all*). Every time it returns
    // Poll::Pending, we assert that it's queued new calls to |fetch| during this
    // iteration, and we resolve each of those response futures. Eventually, it
    // returns Poll::Ready when it's fully walked the source fixture, and at that
    // point we unwrap and return the loaded Tables.
    //
    // Note that the use of BTreeMap above is significant: it means that we resolve
    // those call futures in a stable (sorted) order, and that's the order with which
    // the future will process resolved responses on it's next poll(). It makes the
    // whole mess fully deterministic.

    let mut fut = loader
        .load_resource(
            source::Scope::new(&root),
            &root,
            source::ContentType::CatalogSpec,
        )
        .boxed_local();

    let waker = futures::task::noop_waker();
    let mut ctx = std::task::Context::from_waker(&waker);

    loop {
        match fut.poll_unpin(&mut ctx) {
            Poll::Ready(()) => {
                std::mem::drop(fut);
                return loader.into_tables();
            }
            Poll::Pending if fetches.borrow().is_empty() => {
                // Note the future can return Pending *only because* it's blocked
                // waiting for one or more |fetch| fixtures above to resolve.
                panic!("future is pending, but started no fetches")
            }
            Poll::Pending => {
                for (url, tx) in fetches.borrow_mut().split_off("") {
                    match fixtures.get(&url) {
                        Some(value) => tx.send(Ok(value.to_string().as_bytes().into())),
                        None => tx.send(Err("fixture not found".into())),
                    }
                    .unwrap();
                }
            }
        }
    }
}
