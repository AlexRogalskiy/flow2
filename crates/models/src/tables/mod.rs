use crate::source;
use crate::validation;

#[macro_use]
mod macros;
use macros::*;

pub use macros::{load_tables, persist_tables, Table, TableObj, TableRow};

tables!(
    table Fetches (row Fetch, sql "fetches") {
        resource: url::Url,
    }

    table Resources (row Resource, sql "resources") {
        resource: url::Url,
        content_type: source::ContentType,
        content: Vec<u8>,
    }

    table Imports (row Import, sql "imports") {
        scope: url::Url,
        from_resource: url::Url,
        to_resource: url::Url,
    }

    table NodeJSDependencies (row NodeJSDependency, sql "node_dependencies") {
        scope: url::Url,
        package: String,
        version: String,
    }

    table Collections (row Collection, sql "collections") {
        scope: url::Url,
        collection: source::CollectionName,
        schema: url::Url,
        key: source::CompositeKey,
        store_endpoint: source::EndpointName,
        store_patch_config: serde_json::Value,
    }

    table Projections (row Projection, sql "projections") {
        scope: url::Url,
        collection: source::CollectionName,
        field: String,
        location: source::JsonPointer,
        partition: bool,
        user_provided: bool,
    }

    table Derivations (row Derivation, sql "derivations") {
        scope: url::Url,
        derivation: source::CollectionName,
        register_schema: url::Url,
        register_initial: serde_json::Value,
    }

    table Transforms (row Transform, sql "transforms") {
        scope: url::Url,
        transform: source::TransformName,
        derivation: source::CollectionName,
        source_collection: source::CollectionName,
        source_partitions: Option<source::PartitionSelector>,
        source_schema: Option<url::Url>,
        shuffle_key: Option<source::CompositeKey>,
        shuffle_lambda: Option<source::Lambda>,
        shuffle_hash: source::ShuffleHash,
        read_delay_seconds: Option<u32>,
        priority: u32,
        update_lambda: Option<source::Lambda>,
        publish_lambda: Option<source::Lambda>,
    }

    table Endpoints (row Endpoint, sql "endpoints") {
        scope: url::Url,
        endpoint: source::EndpointName,
        endpoint_type: source::EndpointType,
        base_config: serde_json::Value,
    }

    table Captures (row Capture, sql "captures") {
        scope: url::Url,
        capture: source::CaptureName,
        collection: source::CollectionName,
        allow_push: bool,
        endpoint: Option<source::EndpointName>,
        patch_config: serde_json::Value,
    }

    table Materializations (row Materialization, sql "materializations") {
        scope: url::Url,
        materialization: source::MaterializationName,
        collection: source::CollectionName,
        endpoint: source::EndpointName,
        patch_config: serde_json::Value,
        field_selector: source::MaterializationFields,
    }

    table TestSteps (row TestStep, sql "test_steps") {
        scope: url::Url,
        test: source::TestName,
        step_index: u32,
        step: source::TestStep,
    }

    table SchemaDocs (row SchemaDoc, sql "schema_docs") {
        schema: url::Url,
        dom: serde_json::Value,
    }

    table Inferences (row Inference, sql "inferences") {
        schema: url::Url,
        location: source::JsonPointer,
        spec: protocol::flow::Inference,
    }

    table BuiltCollections (row BuiltCollection, sql "built_collections") {
        scope: url::Url,
        collection: source::CollectionName,
        spec: protocol::flow::CollectionSpec,
    }

    table BuiltMaterializations (row BuiltMaterialization, sql "built_materializations") {
        scope: url::Url,
        materialization: source::MaterializationName,
        collection: source::CollectionName,
        endpoint_config: serde_json::Value,
        field_selection: protocol::materialize::FieldSelection,
    }

    table Errors (row Error, sql "errors") {
        scope: url::Url,
        error: anyhow::Error,
    }
);

// macros::SQLType implementations for table columns.

primitive_sql_types!(
    String => "TEXT",
    url::Url => "TEXT",
    Vec<u8> => "BLOB",
    bool => "BOOLEAN",
    u32 => "INTEGER",
);

string_wrapper_types!(
    source::CaptureName,
    source::CollectionName,
    source::EndpointName,
    source::JsonPointer,
    source::MaterializationName,
    source::TestName,
    source::TransformName,
);

json_sql_types!(
    protocol::flow::CollectionSpec,
    protocol::flow::Inference,
    protocol::flow::shuffle::Hash,
    protocol::materialize::FieldSelection,
    serde_json::Value,
    source::CompositeKey,
    source::ContentType,
    source::EndpointType,
    source::Lambda,
    source::MaterializationFields,
    source::PartitionSelector,
    source::TestStep,
);

// Additional bespoke SQLType implementations for types that require extra help.

impl SQLType for anyhow::Error {
    fn sql_type() -> &'static str {
        "TEXT"
    }
    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput<'_>> {
        Ok(format!("{:?}", self).into())
    }
    fn column_result(value: rusqlite::types::ValueRef<'_>) -> rusqlite::types::FromSqlResult<Self> {
        Ok(anyhow::anyhow!(String::column_result(value)?))
    }
}

impl Errors {
    pub fn push_validation(&mut self, scope: &url::Url, err: validation::Error) {
        self.push_row(scope, anyhow::anyhow!(err))
    }
}
