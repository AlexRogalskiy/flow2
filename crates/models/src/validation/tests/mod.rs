use crate::source;
use crate::tables;
use crate::validation;

use futures::{future::LocalBoxFuture, FutureExt};
use lazy_static::lazy_static;
use protocol::materialize;
use serde_json::Value;
use std::collections::HashMap;

lazy_static! {
    static ref GOLDEN: Value = serde_yaml::from_slice(include_bytes!("model.yaml")).unwrap();
}

#[test]
fn test_golden_all_visits() {
    let (source_tables, built_tables) = run_test(GOLDEN.clone());
    insta::assert_debug_snapshot!(built_tables);
    insta::assert_debug_snapshot!(source_tables);
}

#[test]
fn test_database_round_trip() {
    let (source_tables, built_tables) = run_test(GOLDEN.clone());

    // Round-trip source and built tables through the database, verifying equality.
    let db = rusqlite::Connection::open(":memory:").unwrap();
    tables::persist_tables(&db, &source_tables.as_tables()).unwrap();
    let mut restored_tables = source::Tables::default();
    tables::load_tables(&db, restored_tables.as_tables_mut().as_mut_slice()).unwrap();

    assert_eq!(
        format!("{:#?}", source_tables),
        format!("{:#?}", restored_tables)
    );

    let db = rusqlite::Connection::open(":memory:").unwrap();
    tables::persist_tables(&db, &built_tables.as_tables()).unwrap();
    let mut restored_tables = validation::Tables::default();
    tables::load_tables(&db, restored_tables.as_tables_mut().as_mut_slice()).unwrap();

    assert_eq!(
        format!("{:#?}", built_tables),
        format!("{:#?}", restored_tables)
    );
}

#[test]
fn test_golden_error() {
    run_test_errors(&GOLDEN, "{}");
}

#[test]
fn test_invalid_collection_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  collections:
    good: &spec
      schema: test://int-string.schema
      key: [/int]
      store: { name: storeEndpoint }

    "": *spec
    bad name: *spec
    bad!name: *spec

    # We require a sequence of non-empty tokens, separated by exactly one '/'.
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string & others.
    testing: *spec

    # Illegal duplicates under naming collation.
    testing/int-sTRinG: *spec
    testing/Int-Halve: *spec
"#,
    );
}

#[test]
fn test_invalid_partition_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      projections:
        good: &spec
          location: /bit
          partition: true

        "": *spec
        inv alid: *spec
        inv!alid: *spec
        inv/alid: *spec

        # Illegal duplicates under collation.
        INT: /int
        bIt: /bit

        # Attempt to re-map a canonical projection.
        str: /int
"#,
    );
}

#[test]
fn test_invalid_transform_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          good: &spec
            source:
              name: testing/int-string
            publish:
              nodeJS: "something something"

          "": *spec
          inv alid: *spec
          inv!alid: *spec
          inv/alid: *spec

          # Illegal duplicate under collation.
          reVeRsEIntString: *spec
"#,
    );
}

#[test]
fn test_invalid_endpoint_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  import:
    - test://more-endpoints
test://more-endpoints:
  endpoints:
    good: &spec
      s3:
        bucket: a-bucket
        prefix: and-prefix

    "": *spec
    inv alid: *spec
    inv!alid: *spec
    inv/alid: *spec

    # Illegal duplicates under collation.
    StoreEndpoinT: *spec
    CAPtUReEndpoINT: *spec
"#,
    );
}

#[test]
fn test_invalid_capture_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  captures:
    good: &spec
      target:
        name: testing/int-string
      pushAPI: {}

    "": *spec
    bad name: *spec
    bad!name: *spec
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string/pull & push.
    testing: *spec

    # Illegal duplicates under naming collation.
    testing/int-sTRinG/pUll: *spec
    testing/Int-strINg/PuSH: *spec
"#,
    );
}

#[test]
fn test_invalid_materialization_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  materializations:
    good: &spec
      source:
        name: testing/int-string
      endpoint:
        name: materializeEndpoint
        config: { fixture: one }

    "": *spec
    bad name: *spec
    bad!name: *spec
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string.
    testing: *spec

    # Illegal duplicate under naming collation.
    testing/int-sTRinG: *spec
"#,
    );
}

#[test]
fn test_transform_source_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source: { name: testinG/Int-String }
          halveSelf:
            source: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_capture_target_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  captures:
    testing/int-string/pull:
      target: { name: testiNg/int-strinK }
    testing/int-string/push:
      target: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_capture_endpoint_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  captures:
    testing/int-string/pull:
      endpoint: { name: CaptureEndpoit }
    testing/int-string/push:
      endpoint: { name: wildlyOffName }
      pushAPI: null
"#,
    );
}

#[test]
fn test_capture_endpoint_wrong_type() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  endpoints:
    captureEndpoint:
      s3: null
      postgres:
        host: a-host
        user: a-user
        password: a-password
"#,
    );
}

#[test]
fn test_capture_target_is_derivation_and_missing_imports() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  import: null
  captures:
    testing/int-string/pull:
      target: { name: testing/int-reverse }
"#,
    );
}

#[test]
fn test_capture_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  captures:
    testing/int-string/another/pull:
      target:
        name: testing/int-string
      endpoint: { name: captureEndpoint }
"#,
    );
}

#[test]
fn test_collection_store_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      store: { name: sToreEndpoin }

test://int-reverse:
  collections:
    testing/int-reverse:
      store: { name: WildlyOffName }
"#,
    );
}

#[test]
fn test_use_without_import() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  import: [] # Clear.
  collections:
    testing/int-string:
      store: { name: s3WithoutImport }

test://int-reverse:
  import: [] # Clear.
  endpoints:
    s3WithoutImport:
      s3:
        bucket: a-bucket
        prefix: and-prefix
"#,
    );
}

#[test]
fn test_schema_fragment_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      schema: test://int-string.schema#/not/found

test://int-string-materialization:
  materializations:
    testing/int-string: null # Omit downstream errors.

test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source:
              schema: test://int-string-len.schema#/not/found
"#,
    );
}

#[test]
fn test_keyed_location_wrong_type() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string.schema:
  properties:
    int: { type: [number, object] }
"#,
    );
}

#[test]
fn test_unknown_locations() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      key: [/int, /unknown/key]
      projections:
        Unknown: /unknown/projection

test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/len, /int, /unknown/shuffle]
"#,
    );
}

#[test]
fn test_shuffle_key_length_mismatch() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/len]
          halveSelf:
            shuffle:
              key: [/len, /int]
"#,
    );
}

#[test]
fn test_shuffle_key_types_mismatch() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/int, /str]
          halveSelf:
            shuffle:
              key: [/str, /int]
"#,
    );
}

#[test]
fn test_shuffle_key_empty() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            shuffle: {key: []}
"#,
    );
}

#[test]
fn test_partition_selections() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source:
              partitions:
                include:
                  bit: [true, 42, ""]
                  Int: [15, true]
                  Unknown: ["whoops"]
                exclude:
                  bit: [false, "a string"]
                  Int: [false, "", 16]
                  AlsoUnknown: ["whoops"]
"#,
    );
}

#[test]
fn test_redundant_source_schema_and_shuffle() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            source:
              name: testing/int-string
              schema: test://int-string.schema
            shuffle:
              key: [/int]
"#,
    );
}

#[test]
fn test_incompatible_collection_store() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  endpoints:
    storeEndpoint:
      s3: null
      postgres:
        host: a-host
        user: a-user
        password: a-password
"#,
    );
}

#[test]
fn test_must_have_update_or_publish() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            publish: null
            update: null
"#,
    );
}

#[test]
fn test_invalid_initial_register() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        register:
          initial: "should be an integer"
"#,
    );
}

#[test]
fn test_shape_inspections() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-len.schema:
  properties:
    int:
      reduce: { strategy: set }
    str:
      reduce: { strategy: sum }
"#,
    );
}

#[test]
fn test_schema_reference_verification() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-len.schema:
  $ref: test://int-string.schema#/whoops
"#,
    );
}

#[test]
fn test_materialization_source_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      source: { name: testiNg/int-strinK }

test://int-halve-materialization:
  materializations:
    testing/int-halve:
      source: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_materialization_endpoint_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      endpoint: { name: MaterializeEndpoit }

test://int-halve-materialization:
  materializations:
    testing/int-halve:
      endpoint: { name: wildlyOffName }
"#,
    );
}

#[test]
fn test_materialization_field_errors() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve-materialization:
  materializations:
    testing/int-halve:
      fields:
        include:
          int: {} # Include and exclude.
          biT: {} # Unknown.
        exclude:
          - BiTT # Unknown.
          - WildlyOffName # Also unknown.
          - int
"#,
    );
}

#[test]
fn test_materialization_driver_returns_error() {
    run_test_errors(
        &GOLDEN,
        r#"
driver:
  materializations:
    one:
      constraints: null
      error: "A driver error!"
"#,
    );
}

#[test]
fn test_materialization_driver_unknown_constraint() {
    run_test_errors(
        &GOLDEN,
        r#"
driver:
  materializations:
    one:
      constraints:
        str: { type: 99, reason: "whoops" }
"#,
    );
}

#[test]
fn test_materialization_driver_conflicts() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      fields:
        include:
          str: {}
        exclude:
          - bit
          - Int
driver:
  materializations:
    one:
      constraints:
        flow_document: { type: 1, reason: "location required" }
        Int: { type: 1, reason: "location required" }

        int: { type: 5, reason: "field unsatisfiable" }
        str: { type: 4, reason: "field forbidden" }
        bit: { type: 0, reason: "field required" }
        Unknown: { type: 0, reason: "whoops" }
"#,
    );
}

#[test]
fn test_test_step_unknown_collection() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - ingest:
          collection: testinG/Int-strin
          documents: []
      - verify:
          collection: wildly/Off/Name
          documents: []
"#,
    );
}

#[test]
fn test_test_step_ingest_schema_error() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - ingest:
          collection: testing/int-string
          documents:
            - {int: 42, str_whoops: "string A", bit: true}
            - {int: 52, str_whoops: "string B", bit: true}
"#,
    );
}

#[test]
fn test_test_step_verify_key_order() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - verify:
          collection: testing/int-string
          documents: [{int: 52}, {int: 62}, {int: 42}]
"#,
    );
}

#[test]
fn test_test_step_verify_selector() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - verify:
          collection: testing/int-string
          documents: []
          partitions:
            include:
              bit: [true, 42, ""]
              Int: [15, true]
              Unknown: ["whoops"]
            exclude:
              bit: [false, "a string"]
              Int: [false, "", 16]
              AlsoUnknown: ["whoops"]
"#,
    );
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
struct MockDriverCalls {
    materializations: HashMap<String, MockMaterializationValidateCall>,
}

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
struct MockMaterializationValidateCall {
    endpoint: source::EndpointType,
    config: serde_json::Value,
    #[serde(default)]
    constraints: HashMap<String, materialize::Constraint>,
    #[serde(default)]
    error: Option<String>,
}

impl validation::Drivers for MockDriverCalls {
    fn validate_materialization(
        &self,
        endpoint_type: source::EndpointType,
        endpoint_config: serde_json::Value,
        _request: materialize::ValidateRequest,
    ) -> LocalBoxFuture<Result<materialize::ValidateResponse, validation::BoxError>> {
        async move {
            for (_key, call) in &self.materializations {
                if (call.endpoint, &call.config) != (endpoint_type, &endpoint_config) {
                    continue;
                }

                if let Some(err) = &call.error {
                    return Err(err.as_str().into());
                } else {
                    return Ok(materialize::ValidateResponse {
                        constraints: call.constraints.clone(),
                    });
                }
            }
            return Err("driver fixture not found".into());
        }
        .boxed_local()
    }
}

fn run_test(mut fixture: Value) -> (source::Tables, validation::Tables) {
    // Extract out driver mock call fixtures.
    let mock_calls = fixture
        .get_mut("driver")
        .map(|d| d.take())
        .unwrap_or_default();
    let mock_calls: MockDriverCalls = serde_json::from_value(mock_calls).unwrap();

    let source_tables = source::scenarios::evaluate_fixtures(Default::default(), &fixture);
    if !source_tables.errors.is_empty() {
        eprint!("{:?}", &source_tables);
        panic!("unexpected fixture load error");
    }
    let built_tables = validation::validate(&mock_calls, &source_tables);

    (source_tables, built_tables)
}

fn run_test_errors(fixture: &Value, patch: &str) {
    let mut fixture = fixture.clone();
    let patch: Value = serde_yaml::from_str(patch).unwrap();
    json_patch::merge(&mut fixture, &patch);

    let (_, built_tables) = run_test(fixture);
    insta::assert_debug_snapshot!(built_tables.errors);
}
