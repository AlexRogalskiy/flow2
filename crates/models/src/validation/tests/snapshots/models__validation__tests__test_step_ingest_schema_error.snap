---
source: crates/models/src/validation/tests/mod.rs
expression: tables.errors
---
[
    Error {
        scope: test://int-string-tests#/tests/A%20Test/0,
        error: test ingest document is invalid against the collection schema: {
          "document": {
            "bit": true,
            "int": 42,
            "str_whoops": "string A"
          },
          "basic_output": {
            "errors": [
              {
                "absoluteKeywordLocation": "test://int-string.schema",
                "error": "Invalid(Required { props: [\"int\", \"str\", \"bit\"], props_interned: 7 })",
                "instanceLocation": "",
                "keywordLocation": "#"
              }
            ],
            "valid": false
          }
        },
    },
    Error {
        scope: test://int-string-tests#/tests/A%20Test/0,
        error: test ingest document is invalid against the collection schema: {
          "document": {
            "bit": true,
            "int": 52,
            "str_whoops": "string B"
          },
          "basic_output": {
            "errors": [
              {
                "absoluteKeywordLocation": "test://int-string.schema",
                "error": "Invalid(Required { props: [\"int\", \"str\", \"bit\"], props_interned: 7 })",
                "instanceLocation": "",
                "keywordLocation": "#"
              }
            ],
            "valid": false
          }
        },
    },
]
